lib/model/autocomplete.dart

```diff
@@ -23,26 +24,126 @@ extension Autocomplete on ComposeContentController {
+          textEditingValue: value);
       }
-      if (selection.start < position) {
-        // See comment about [TextSelection.isCollapsed] above.
-        return null;
+
+      if (textUntilCursor[position] == '#') {
+        final match = channelMentionAutocompleteMarkerRegex.matchAsPrefix(textUntilCursor, position);
+        if (match == null) {
+          continue;
+        }
+        if (selection.start < position) {
+          // See comment about [TextSelection.isCollapsed] above.
+          return null;
+        }
+        return AutocompleteIntent(
+          syntaxStart: position,
+          query: ChannelMentionAutocompleteQuery(match[1]!),
+          textEditingValue: value);
       }
-      return AutocompleteIntent(
-        syntaxStart: position,
-        query: MentionAutocompleteQuery(match[2]!, silent: match[1]! == '_'),
-        textEditingValue: value);
     }
     return null;
   }
 }
 
+extension ComposeTopicAutocomplete on ComposeTopicController {
+  AutocompleteIntent<TopicAutocompleteQuery>? autocompleteIntent() {
+    return AutocompleteIntent(
+      syntaxStart: 0,
+      query: TopicAutocompleteQuery(value.text),
+      textEditingValue: value);
+  }
+}
+
+const wordBoundaryChars = " _/-";
+
+final class _TriageRawData<T> {
+  final List<T> exactMatches;
+  final List<T> beginsWithCaseSensitiveMatches;
+  final List<T> beginsWithCaseInsensitiveMatches;
+  final List<T> wordBoundaryMatches;
+  final List<T> noMatches;
+
+  _TriageRawData({
+    required this.exactMatches,
+    required this.beginsWithCaseSensitiveMatches,
+    required this.beginsWithCaseInsensitiveMatches,
+    required this.wordBoundaryMatches,
+    required this.noMatches});
+}
+
+_TriageRawData<T> _triageRaw<T>({
+  required String query,
+  required Iterable<T> objects,
+  required String Function(T) getItem}) {
+  final exactMatches = <T>[];
+  final beginsWithCaseSensitiveMatches = <T>[];
+  final beginsWithCaseInsensitiveMatches = <T>[];
+  final wordBoundaryMatches = <T>[];
+  final noMatches = <T>[];
+  final lowerQuery = query.toLowerCase();
+
+  for (var object in objects) {
+    final item = getItem(object);
+    final lowerItem = item.toLowerCase();
+
+    if (lowerQuery == lowerItem) {
+      exactMatches.add(object);
+    } else if (item.startsWith(query)) {
+      beginsWithCaseSensitiveMatches.add(object);
+    } else if (lowerItem.startsWith(lowerQuery)) {
+      beginsWithCaseInsensitiveMatches.add(object);
+    } else if (RegExp('[$wordBoundaryChars]${RegExp.escape(lowerQuery)}').hasMatch(lowerItem)) {
+      wordBoundaryMatches.add(object);
+    } else {
+      noMatches.add(object);
+    }
+  }
+  return _TriageRawData(exactMatches: exactMatches,
+    beginsWithCaseSensitiveMatches: beginsWithCaseSensitiveMatches,
+    beginsWithCaseInsensitiveMatches: beginsWithCaseInsensitiveMatches,
+    wordBoundaryMatches: wordBoundaryMatches,
+    noMatches: noMatches);
+}
+
+({List<T> matches, List<T> rest}) _triage<T> ({
```

Comment:
Because this function (and its helper `_triageRaw`) are designed to closely mimic particular pieces of web's logic, they should get comments linking to the corresponding web code.

That helps a reviewer compare to the web code to see what differences there might be. It also helps in the future — particularly if that link is a permalink, which it should be — for a reader to compare against what's changed in web since then, and see if changes should be copied over to here.